nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,804 @@
+                    continue;
+                }
+                self.radio.events_end.reset();
+            }
+            Event::PhyEnd => {
+                while self
+                    .radio
+                    .events_phyend
+                    .read()
+                    .events_phyend()
+                    .bit_is_clear()
+                {
+                    continue;
+                }
+                self.radio.events_phyend.reset();
+            }
+        }
+    }
+
+    /// Waits until the radio state matches the given `state`
+    fn wait_for_state_a(&self, state: STATE_A) {
+        while self.radio.state.read().state() != state {
+            continue;
+        }
+    }
+}
+
+/// Error
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Error {
+    /// Incorrect CRC
+    Crc(u16),
+    /// Timeout
+    Timeout,
+}
+
+/// Driver state
+///
+/// After, or at the start of, any method call the RADIO will be in one of these states
+// This is a subset of the STATE_A enum
+#[derive(Copy, Clone, PartialEq)]
+enum State {
+    Disabled,
+    RxIdle,
+    TxIdle,
+}
+
+/// NOTE must be followed by a volatile write operation
+fn dma_start_fence() {
+    atomic::compiler_fence(Ordering::Release);
+}
+
+/// NOTE must be preceded by a volatile read operation
+fn dma_end_fence() {
+    atomic::compiler_fence(Ordering::Acquire);
+}
+
+enum Event {
+    End,
+    PhyEnd,
+}
+
+/// An IEEE 802.15.4 packet
+///
+/// This `Packet` is a PHY layer packet. It's made up of the physical header (PHR) and the PSDU
+/// (PHY service data unit). The PSDU of this `Packet` will always include the MAC level CRC, AKA
+/// the FCS (Frame Control Sequence) -- the CRC is fully computed in hardware and automatically
+/// appended on transmission and verified on reception.
+///
+/// The API lets users modify the usable part (not the CRC) of the PSDU via the `deref` and
+/// `copy_from_slice` methods. These methods will automatically update the PHR.
+///
+/// See figure 119 in the Product Specification of the nRF52840 for more details
+pub struct Packet {
+    buffer: [u8; Self::SIZE],
+}
+
+// See figure 124 in nRF52840-PS
+impl Packet {
+    // for indexing purposes
+    const PHY_HDR: usize = 0;
+    const DATA: RangeFrom<usize> = 1..;
+
+    /// Maximum amount of usable payload (CRC excluded) a single packet can contain, in bytes
+    pub const CAPACITY: u8 = 125;
+    const CRC: u8 = 2; // size of the CRC, which is *never* copied to / from RAM
+    const MAX_PSDU_LEN: u8 = Self::CAPACITY + Self::CRC;
+    const SIZE: usize = 1 /* PHR */ + Self::MAX_PSDU_LEN as usize;
+
+    /// Returns an empty packet (length = 0)
+    // XXX I believe that be making this *not* `const` it is not possible to place a `Packet` in
+    // `.rodata` (modulo unsafe `#[link_section]` shenanigans) thus it is not necessary to check the
+    // address of packet in `Radio.{send,recv}` (EasyDMA can only operate on RAM addresses)
```

Comment:
Hmm, I think I'd prefer not to rely on this. Instead, `recv` and `send` could both take a `&mut Packet`, which ensures that it's in writeable RAM (other modules in the HAL already skip the address check when they deal with mutable data).